### ISSUE TITLE
feat: Make dart_frog_gen great again.

### ISSUE THREAD

**Issue by renancaraujo** [VeryGoodOpenSource/dart_frog#1016]
### tl;dr

This is a proposal for a major refactor on how code is organized within the Dart Frog tooling with the goal of promoting greater composability.

###  **The problem**

One can identify Dart Frog as not only. package, but an SDK composed for the runtime package (`dart_frog`), its companion first-party packages (`dart_frog auth` and `dart_frog_websocket`, for instance), and most importantly, its tooling, which is the subject of this proposal. 

The Dart Frog tooling has its own elements of composition, with two major frontends, the VS Code extension and the `dart_frog_cli`. 

As the backend for the tooling, there is `dart_frog_gen`, which is responsible for reading the file system under a project's `routes` and generating a route config from it. 

The problem is that we are under-using dart_frog_gen and overusing dart_frog_cli to include all the code generation algorithms. 

In the following excalidraw, we can see the current situation:

[![currrent situation](https://github.com/VeryGoodOpenSource/dart_frog/assets/6718144/22c9b139-b15f-41e2-a937-4b3a4b73d6f4)](https://excalidraw.com/#json=7NIaY0F8MdrkgfqB17cGU,aWDiWtjCohjj4z17ICfNrw). 

First and third-party tooling has to depend on the CLI (as a library) to access features such as:

- Daemon messages structure (Affects use cases such as #1013)
- Dev server runner
- Route config watching

###  **The proposed solution**

Move the backend code to dart_frog_gen. 

[![after proposal situation](https://github.com/VeryGoodOpenSource/dart_frog/assets/6718144/9702a21d-eb01-4611-bab1-455a3689eb6c)](https://excalidraw.com/#json=i_wkHnN6HG3N84hNG9FW-,tGCAShodGSZZI12ZMyjDYQ)

This would require the gen package to include the following features:
- dev server runner
- runtime compatibility checks
- daemon protocol

### Gains

This enables the creation of packages to depend on `dart_frog_gen` directly without having to rely on the user-installed CLI. 
Examples of issues and packages that could :
- A client package for the daemon
- #474
- #152

**Comment by alestiago** [comment]
This refactor can be broken down into many parts. For example, #1161 contributes to this refactor. Some of these mentioned parts might be more sensible to refactor than others.

**Comment by dgaedcke** [comment]
Curious if OpenApi (https://github.com/VeryGoodOpenSource/dart_frog/issues/152) will be included in this refactor?  And if so, any ETA?
